designs/2019-worker-threads/README.md

```diff
@@ -0,0 +1,293 @@
+│    ├ ...
+│    └ 🔁 Read-lint-report loop
+├ ⏩ Worker thread
+│    ├ 🔁 Read-lint-report loop
+│    ├ ...
+│    └ 🔁 Read-lint-report loop
+├ ...
+└ ⏩ Worker thread
+      ├ 🔁 Read-lint-report loop
+      ├ ...
+      └ 🔁 Read-lint-report loop
+```
+
+#### Loading Config
+
+On the other hand, this RFC doesn't change the logic that loads config files. It's still synchronous. Because the logic is large and the configs are cached in most cases.
+
+We can improve config loading by async in the future, but it's out of the scope of this RFC.
+
+#### Constants
+
+This RFC contains two constants.
+
+- `128` ... the denominator to estimate the proper `concurrency` option value.
+- `16` ... concurrency for non-blocking IO.
+
+Those values are changeable by measuring performance in real.
+
+#### Performance
+
+Please try the PoC.
+
+```
+npm install github:eslint/eslint#very-rough-worker-threads-poc
+```
+
+This section is an instance of performance measurement. The worker threads make ESLint over 2x faster if there are many files.
+
+- Processor: Intel Core i7-8700 CPU (12 processors)
+- RAM: 16.0 GB
+- OS: Windows 10 Pro 64bit
+- Node.js: 12.11.0
+
+```
+$ eslint lib tests/lib --concurrency 1
+Number of files: 697
+Lint files in the main thread.
+Linting complete in: 19636ms
+
+$ eslint lib tests/lib --concurrency auto
+Number of files: 697
+Lint files in 6 worker threads.
+Linting complete in: 7486ms
+
+$ eslint lib --concurrency 1
+Number of files: 368
+Lint files in the main thread.
+Linting complete in: 8442ms
+
+$ eslint lib --concurrency auto
+Number of files: 368
+Lint files in 3 worker threads.
+Linting complete in: 4672ms
+```
+
+If forced to use workers with a few files, it's slower.
+
+```
+$ eslint lib/cli-engine --concurrency auto
+Number of files: 28
+Lint files in the main thread.
+Linting complete in: 1521ms
+
+$ eslint lib/cli-engine --concurrency 2
+Number of files: 28
+Lint files in 2 worker threads.
+Linting complete in: 2054ms
+```
+
+### [STEP 2] Changing the default of `concurrency` option to `"auto"`
+
+This change needs a major release.
+
+ESLint gets using worker threads by default if target files are many.
+
+## Documentation
+
+- The "[Command Line Interface](https://eslint.org/docs/user-guide/command-line-interface)" page should describe new `--concurrency` option.
+- The "[Node.js API](https://eslint.org/docs/developer-guide/nodejs-api)" page should describe new `concurrency` option.
+- When step 2, it needs an entry of the migration guide because of a breaking change.
+
+## Drawbacks
+
+This feature introduces complexity for parallelism. It will be an additional cost to maintain our codebase.
+
+This feature might not fit `eslint-plugin-import` and `@typescript-eslint`. Those have to parse the entire codebase to check types, but workers cannot share the parsed information. Every worker has to parse the entire codebase individually. But linting in parallel is faster than single-thread in spite of the cost of duplicate works.
```

Comment:
You mentioned yourself earlier that `SharedArrayBuffer` could be used to share data and @ljharb mentioned that one could maybe add an initial first step in the main thread to allow for plugins like these to handle some of these special cases – maybe add both of those thoughts here as well to get a more nuanced view of the compatibility "risk"?

One could also imagine a scenario like when Firefox moved to a multi-thread model: That plugins can indicate themselves as incompatible with a multi-threaded ESLint and that ESLint in that case disables multiple threads even in the Step 2 scenario (and a Step 3 scenario could maybe be added that eventually deprecates and removes that possibility, just like Firefox did)